api/metrics/v1/rules.go

```diff
@@ -0,0 +1,239 @@
+			level.Warn(logger).Log("msg", msg)
+			http.Error(w, msg, http.StatusInternalServerError)
+			return
+		}
+
+		rules, err := lister.ListRuleGroups(r.Context(), tenant)
+		if err != nil {
+			msg := "failed to list rules"
+			level.Debug(logger).Log("msg", msg, "err", err)
+			http.Error(w, msg, http.StatusInternalServerError)
+			return
+		}
+
+		for i := range rules.Groups {
+			for j := range rules.Groups[i].Rules {
+				if rules.Groups[i].Rules[j].Labels == nil {
+					rules.Groups[i].Rules[j].Labels = make(map[string]string)
+				}
+				rules.Groups[i].Rules[j].Labels[label] = id
+			}
+		}
+
+		bytes, err := yaml.Marshal(rules)
+		if err != nil {
+			msg := "failed to marshal rules"
+			level.Debug(logger).Log("msg", msg, "err", err)
+			http.Error(w, msg, http.StatusInternalServerError)
+			return
+		}
+
+		_, _ = w.Write(bytes)
+	}
+}
+
+type RulesGetter interface {
+	GetRules(ctx context.Context, tenant, name string) (RuleGroup, error)
+}
+
+func getRuleHandler(logger log.Logger, repository RulesGetter, label string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		tenant, ok := authentication.GetTenant(r.Context())
+		if !ok {
+			http.Error(w, "failed to get tenant", http.StatusInternalServerError)
+			return
+		}
+
+		id, ok := authentication.GetTenantID(r.Context())
+		if !ok {
+			const msg = "error finding tenant ID"
+			level.Warn(logger).Log("msg", msg)
+			http.Error(w, msg, http.StatusInternalServerError)
+			return
+		}
+
+		name := chi.URLParam(r, "name")
+
+		rules, err := repository.GetRules(r.Context(), tenant, name)
+		if err == ErrRuleNotFound {
+			msg := "rule not found"
+			level.Debug(logger).Log("msg", msg)
+			http.Error(w, msg, http.StatusNotFound)
+			return
+		}
+		if err != nil {
+			msg := "failed to get rules"
+			level.Warn(logger).Log("msg", msg, "err", err)
+			http.Error(w, msg, http.StatusInternalServerError)
+			return
+		}
+
+		for i := range rules.Rules {
+			if rules.Rules[i].Labels == nil {
+				rules.Rules[i].Labels = make(map[string]string)
+			}
+			rules.Rules[i].Labels[label] = id
+		}
+
+		bytes, err := yaml.Marshal(rules)
+		if err != nil {
+			msg := "failed to marshal rules"
+			level.Warn(logger).Log("msg", msg, "err", err)
+			http.Error(w, msg, http.StatusInternalServerError)
+			return
+		}
+
+		_, _ = w.Write(bytes)
+	}
+}
+
+type RulesWriter interface {
+	CreateRule(ctx context.Context, tenant, name string, interval int64, rules []byte) error
+	UpdateRule(ctx context.Context, tenant, name string, interval int64, rules []byte) error
+	DeleteRule(ctx context.Context, tenant, name string) error
+}
+
+func writeRuleHandler(logger log.Logger, repository RulesWriter, label string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		tenant, ok := authentication.GetTenant(r.Context())
+		if !ok {
+			const msg = "failed to get tenant"
+			level.Warn(logger).Log("msg", msg)
+			http.Error(w, msg, http.StatusInternalServerError)
```

Comment:
is an incorrect tenant really an `http.StatusInternalServerError`? I would have thought `http.StatusBadRequest`?

Reply:
wait ignore me - that's it failing to parse from the context :facepalm: yeah deffo 5xx